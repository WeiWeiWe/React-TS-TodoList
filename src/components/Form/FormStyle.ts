import styled from 'styled-components';

export const FormStyle = styled.section`
  .form-fields {
    display: flex;

    .from-input {
      flex: 1;
      font-size: 18px;
      border: none;
      outline: none;
    }

    .from-todo-icon {
      flex: 0 0 25px;
      padding: 9px;
    }

    .from-add-icon {
      flex: 0 0 25px;
      padding: 9px;
      cursor: pointer;
    }
  }
`;
