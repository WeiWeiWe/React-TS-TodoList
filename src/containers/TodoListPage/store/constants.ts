export const CHANGE_INPUT_VALUE = 'CHANGE_INPUT_VALUE';
export const ADD_TODO_LIST = 'ADD_TODO_LIST';
export const DELETE_TODO_LIST = 'DELETE_TODO_LIST';
export const CHANGE_LIST_ITEM_TITLE_VALUE = 'CHANGE_LIST_ITEM_TITLE_VALUE';
export const CHANGE_PROCESS_STATUS = 'CHANGE_PROCESS_STATUS';
export const CHANGE_TAB_STATUS = 'CHANGE_TAB_STATUS';
