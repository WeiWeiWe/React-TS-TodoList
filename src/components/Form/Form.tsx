import React from 'react';
import { Reader, AddCircleSharp } from 'react-ionicons';
import { FormStyle } from './FormStyle';

export interface FormIprops {
  inputValue: string;
  changeInputValue: (e: React.ChangeEvent<HTMLInputElement>) => void;
  addTodoList: () => void;
  keyUpTodoList: (e: React.KeyboardEvent<HTMLInputElement>) => void;
}

export function Form({ inputValue, changeInputValue, addTodoList, keyUpTodoList }: FormIprops) {
  return (
    <FormStyle>
      <div className="global-block-fields form-fields">
        <i className="from-todo-icon">
          <Reader color="rgb(50, 197, 250)" width="35px" height="30px" />
        </i>
        <input
          type="text"
          className="from-input"
          placeholder="Enter your next task"
          value={inputValue}
          onChange={changeInputValue}
          onKeyUp={keyUpTodoList}
          data-testid="form-input"
        />
        <i className="from-add-icon" onClick={addTodoList} data-testid="add-btn">
          <AddCircleSharp cssClasses="add-btn" color="rgba(50, 197, 250)" width="30px" height="30px" />
        </i>
      </div>
    </FormStyle>
  );
}
