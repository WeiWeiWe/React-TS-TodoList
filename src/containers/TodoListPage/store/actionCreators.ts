import * as constants from './constants';
import { ItemTypes, EnumItemProcessTypes } from '../../../types';

interface ChangeInputValueAction {
  type: typeof constants.CHANGE_INPUT_VALUE;
  value: string;
}

interface AddTodoListAction {
  type: typeof constants.ADD_TODO_LIST;
  item: ItemTypes;
}

interface DeleteTodoListAction {
  type: typeof constants.DELETE_TODO_LIST;
  id: string;
}

interface SwitchModifyStatusAction {
  type: typeof constants.SWITCH_MODIFY_STATUS;
  status: boolean;
  id: string;
}

interface ChangeListItemTitleValueAction {
  type: typeof constants.CHANGE_LIST_ITEM_TITLE_VALUE;
  value: string;
  id: string;
}

interface ChangeProcessStatusAction {
  type: typeof constants.CHANGE_PROCESS_STATUS;
  process: string;
  id: string;
}

interface ChangeTabStatusAction {
  type: typeof constants.CHANGE_TAB_STATUS;
  status: EnumItemProcessTypes;
}

export type TodoListActionTypes =
  | ChangeInputValueAction
  | AddTodoListAction
  | DeleteTodoListAction
  | SwitchModifyStatusAction
  | ChangeListItemTitleValueAction
  | ChangeProcessStatusAction
  | ChangeTabStatusAction;

export const changeInputValueActionCreator = (value: string): ChangeInputValueAction => {
  return {
    type: constants.CHANGE_INPUT_VALUE,
    value,
  };
};

export const addTodoListActionCreator = (item: ItemTypes): AddTodoListAction => {
  return {
    type: constants.ADD_TODO_LIST,
    item,
  };
};

export const deleteTodoListActionCreator = (id: string): DeleteTodoListAction => {
  return {
    type: constants.DELETE_TODO_LIST,
    id,
  };
};

export const switchModifyStatusActionCreator = (status: boolean, id: string): SwitchModifyStatusAction => {
  return {
    type: constants.SWITCH_MODIFY_STATUS,
    status,
    id,
  };
};

export const ChangeListItemTitleValueActionCreator = (value: string, id: string): ChangeListItemTitleValueAction => {
  return {
    type: constants.CHANGE_LIST_ITEM_TITLE_VALUE,
    value,
    id,
  };
};

export const ChangeProcessStatusActionCreator = (process: string, id: string): ChangeProcessStatusAction => {
  return {
    type: constants.CHANGE_PROCESS_STATUS,
    process,
    id,
  };
};

export const ChangeTabStatusActionCreator = (status: EnumItemProcessTypes): ChangeTabStatusAction => {
  return {
    type: constants.CHANGE_TAB_STATUS,
    status,
  };
};
