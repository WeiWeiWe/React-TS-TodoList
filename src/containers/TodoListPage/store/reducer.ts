import * as constants from './constants';
import * as actionCreators from './actionCreators';
import { ItemTypes } from '../../../types';

export interface defaultStateTypes {
  inputValue: string;
  list: ItemTypes[];
}

const defaultState: defaultStateTypes = {
  inputValue: '',
  list: [],
};

export default (state = defaultState, action: actionCreators.TodoListActionTypes) => {
  switch (action.type) {
    case constants.CHANGE_INPUT_VALUE:
      return {
        ...state,
        inputValue: action.value,
      };
    case constants.ADD_TODO_LIST:
      return {
        ...state,
        inputValue: '',
        list: [...state.list, action.item],
      };
    case constants.DELETE_TODO_LIST:
      const newList = state.list.filter((item, index) => index !== action.index);
      return {
        ...state,
        list: newList,
      };
    default:
      return state;
  }
};
