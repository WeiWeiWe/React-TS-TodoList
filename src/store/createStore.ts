import { createStore, combineReducers } from 'redux';
import { reducer as todoReducer } from '../containers/TodoListPage/store';

const rootReducer = combineReducers({
  todo: todoReducer,
});

const store = createStore(rootReducer);

export type RootState = ReturnType<typeof store.getState>;

export default store;
