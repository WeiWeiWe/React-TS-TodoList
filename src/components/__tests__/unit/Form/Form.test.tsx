import React from 'react';
import { Provider } from 'react-redux';
import { render, fireEvent } from '@testing-library/react';
import store from '../../../../store/createStore';
import { Form, FormIprops } from '../../../../components/Form/Form';

const testProps: FormIprops = {
  inputValue: '',
  changeInputValue: jest.fn(),
  addTodoList: jest.fn(),
  keyUpTodoList: jest.fn(),
};

const renderComponentWithRedux = () => {
  return (
    <Provider store={store}>
      <Form {...testProps} />
    </Provider>
  );
};

describe('Test Form Component', () => {
  test('should render the component to match the snapshot', () => {
    const wrapper = render(renderComponentWithRedux());
    expect(wrapper).toMatchSnapshot();
  });

  test('render the component should contains input element and add button', () => {
    const { getByTestId } = render(renderComponentWithRedux());
    const inputElem = getByTestId(/form-input/) as HTMLInputElement;
    const addButton = getByTestId(/add-btn/) as HTMLButtonElement;
    expect(inputElem).toBeInTheDocument;
    expect(addButton).toBeInTheDocument;
  });

  test('should input element value is empty initially', () => {
    const { getByTestId } = render(renderComponentWithRedux());
    const inputElem = getByTestId(/form-input/) as HTMLInputElement;
    expect(inputElem.value).toBe('');
  });

  test('When user input, should trigger the changeInputValue function', () => {
    const { getByTestId } = render(renderComponentWithRedux());
    const inputElem = getByTestId(/form-input/) as HTMLInputElement;
    const userInput = 'Buy pencil';

    fireEvent.change(inputElem, { target: { value: userInput } });

    expect(testProps.changeInputValue).toHaveBeenCalled();
  });

  test('When user key up the enter keyboard, should trigger the keyUpTodoList function', () => {
    const { getByTestId } = render(renderComponentWithRedux());
    const inputElem = getByTestId(/form-input/) as HTMLInputElement;

    fireEvent.keyUp(inputElem, { key: 'Enter', code: 'Enter', charCode: 13 });

    expect(testProps.keyUpTodoList).toHaveBeenCalled();
  });

  test('When user click the add button, should trigger the addTodoList function', () => {
    const { getByTestId } = render(renderComponentWithRedux());
    const addButton = getByTestId(/add-btn/) as HTMLButtonElement;

    fireEvent.click(addButton);

    expect(testProps.addTodoList).toHaveBeenCalled();
  });
});
