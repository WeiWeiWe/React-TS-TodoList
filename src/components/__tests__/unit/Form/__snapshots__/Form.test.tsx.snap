// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Form Component should render the component to match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="sc-gsTEea fZJIli"
      >
        <div
          class="global-block-fields form-fields"
        >
          <i
            class="from-todo-icon"
          >
            <span
              class="sc-bdfBQB fLHKPW"
              color="rgb(50, 197, 250)"
              height="30px"
              rotate="0"
              width="35px"
            >
              <svg
                class=""
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                
                <path
                  d="M368 32H144a64.07 64.07 0 00-64 64v320a64.07 64.07 0 0064 64h224a64.07 64.07 0 0064-64V96a64.07 64.07 0 00-64-64zM256 304h-80a16 16 0 010-32h80a16 16 0 010 32zm80-80H176a16 16 0 010-32h160a16 16 0 010 32zm0-80H176a16 16 0 010-32h160a16 16 0 010 32z"
                />
              </svg>
            </span>
          </i>
          <input
            class="from-input"
            data-testid="form-input"
            placeholder="Enter your next task"
            type="text"
            value=""
          />
          <i
            class="from-add-icon"
            data-testid="add-btn"
          >
            <span
              class="sc-bdfBQB ihcyJf"
              color="rgba(50, 197, 250)"
              height="30px"
              rotate="0"
              width="30px"
            >
              <svg
                class="add-btn"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                
                <path
                  d="M256 48C141.31 48 48 141.31 48 256s93.31 208 208 208 208-93.31 208-208S370.69 48 256 48zm96 224h-80v80h-32v-80h-80v-32h80v-80h32v80h80z"
                />
              </svg>
            </span>
          </i>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="sc-gsTEea fZJIli"
    >
      <div
        class="global-block-fields form-fields"
      >
        <i
          class="from-todo-icon"
        >
          <span
            class="sc-bdfBQB fLHKPW"
            color="rgb(50, 197, 250)"
            height="30px"
            rotate="0"
            width="35px"
          >
            <svg
              class=""
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              
              <path
                d="M368 32H144a64.07 64.07 0 00-64 64v320a64.07 64.07 0 0064 64h224a64.07 64.07 0 0064-64V96a64.07 64.07 0 00-64-64zM256 304h-80a16 16 0 010-32h80a16 16 0 010 32zm80-80H176a16 16 0 010-32h160a16 16 0 010 32zm0-80H176a16 16 0 010-32h160a16 16 0 010 32z"
              />
            </svg>
          </span>
        </i>
        <input
          class="from-input"
          data-testid="form-input"
          placeholder="Enter your next task"
          type="text"
          value=""
        />
        <i
          class="from-add-icon"
          data-testid="add-btn"
        >
          <span
            class="sc-bdfBQB ihcyJf"
            color="rgba(50, 197, 250)"
            height="30px"
            rotate="0"
            width="30px"
          >
            <svg
              class="add-btn"
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              
              <path
                d="M256 48C141.31 48 48 141.31 48 256s93.31 208 208 208 208-93.31 208-208S370.69 48 256 48zm96 224h-80v80h-32v-80h-80v-32h80v-80h32v80h80z"
              />
            </svg>
          </span>
        </i>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
